In [38]:
NAME = "Niloy Farhan"
ID = "20101118"
COLLABORATORS_ID = ["21241056", ""]

---

# First import the necessary libraries

In [39]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn import datasets, preprocessing, linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Linear Regression

In [40]:
# Import the diabetes dataset from sklearn's datasets module, save it to the variable data
data = datasets.load_diabetes()

# Split the data into X and y. Print the shape of the two array.
X = data['data']
y = data['target']

### BEGIN SOLUTION
print("Shape of X: {}".format(X.shape))
print("Shape of y: {}".format(y.shape))
### END SOLUTION

Shape of X: (442, 10)
Shape of y: (442,)


## Question 1
How many samples are there in this datasets? How many features are there per sample?

There are 442 samples in the datasets. There are 10 features per sample.

In [41]:
# Display the first 5 rows of the dataset in a pandas dataframe format

### BEGIN SOLUTION
dataframe = pd.DataFrame(X, columns = data['feature_names'])
dataframe.head()
### END SOLUTION

age       sex       bmi  ...        s4        s5        s6
0  0.038076  0.050680  0.061696  ... -0.002592  0.019908 -0.017646
1 -0.001882 -0.044642 -0.051474  ... -0.039493 -0.068330 -0.092204
2  0.085299  0.050680  0.044451  ... -0.002592  0.002864 -0.025930
3 -0.089063 -0.044642 -0.011595  ...  0.034309  0.022692 -0.009362
4  0.005383 -0.044642 -0.036385  ... -0.002592 -0.031991 -0.046641

[5 rows x 10 columns]

In [42]:
# Print the maximum and the minimum of each each feature

### BEGIN SOLUTION
print("Feature wise maximum")
print(dataframe.max())
print("Feature wise minimum")
print(dataframe.min())
### END SOLUTION

Feature wise maximum
age    0.110727
sex    0.050680
bmi    0.170555
bp     0.132044
s1     0.153914
s2     0.198788
s3     0.181179
s4     0.185234
s5     0.133599
s6     0.135612
dtype: float64
Feature wise minimum
age   -0.107226
sex   -0.044642
bmi   -0.090275
bp    -0.112400
s1    -0.126781
s2    -0.115613
s3    -0.102307
s4    -0.076395
s5    -0.126097
s6    -0.137767
dtype: float64


## Question 2
Are the minimums and the maximums close to each other? Hence, do we need any normalization or standardization?

Yes, The minimums and maximums are close to each other. Hence, we don't need any normalization.

In [43]:
# Split the data into training and testing set, with 20% of the data in the testing set. Use random_state = 100
# Then print the shapes of X_train and X_test to see how many samples are on each set

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size= 0.20, random_state= 100) # replace with proper code

### BEGIN SOLUTION
print(" Shape of X_train is {}".format(X_train.shape))
print(" Shape of X_test is {}".format(X_test.shape))
### END SOLUTION

 Shape of X_train is (353, 10)
 Shape of X_test is (89, 10)


In [44]:
# Fit a linear regression model on X_train, y_train
# predict the model's output for X_train and save it to y_predict_train
# predict the model's output for X_test and save it to y_predict_test


### BEGIN SOLUTION
model = linear_model.LinearRegression()
model.fit(X_train, y_train)
y_predict_train = model.predict(X_train)
y_predict_test = model.predict(X_test)
### END SOLUTION

In [45]:
# Calculate a suitable performance metric for the training set for this regression problem
# Calculate a suitable performance metric for the test set for this regression problem
# Print the values

performance_train = np.mean((y_train - y_predict_train)**2)**0.5
performance_test = np.mean((y_test - y_predict_test)**2)**0.5

### BEGIN SOLUTION
print("RMSE on training data=", performance_train)
print("RMSE on testing data=", performance_test)
### END SOLUTION

RMSE on training data= 54.11079192361949
RMSE on testing data= 51.33052418788044


## Question 3
- Which performance metric did you use and why?
- Given the minimum and the maximum of `y` are 25 and 346, respectively, is this a good model for the dataset? Why or why not?
- Suggest some ways (at least one) that might improve the performance metric

I use Root Mean Squared Error to measure the RMSE because it gives us an idea about the average error between actual value and prediction value. 

Given the minimum and the maximum of y are 25 and 346, respectively, This is not a good model. By using RMSE we find out that our mean error of our prediction is 51 in test data. For example, if y = 25 then our predicted value can be between (25-51) = -26 to 25+51 = 76 which can be a really bad prediction flactuating 2-3 times more or less. This is why it is not a good model.

To improve this model, we can try other machine learning algorithms. We can also tweek the hyper parameters. Choosing the right features carefully will also improve this model significantly.


# Classification

In [46]:
# Import the sonar dataset (contained in the sonar_data.csv file)
df = pd.read_csv("sonar_data.csv")

# Split it into X and y
X = df.drop(columns=['y']).values
y = df['y'].values
print("Shape of X -", X.shape)
print("Shape of y -", y.shape)
print("Unique values of y -", np.unique(y))

# Printing the first 5 rows
df.head()


Shape of X - (208, 60)
Shape of y - (208,)
Unique values of y - [0 1]


x1      x2       x3       x4  ...      x58     x59     x60  y
0 -4.7000  2.9275  25.0272  10.4968  ...  23.1764  8.0630  3.0800  0
1 -4.3205  3.3075  26.0232  11.6536  ...  23.1029  8.0364  3.1100  0
2 -4.6070  3.4550  26.6376  12.5992  ...  23.3444  8.0665  3.1950  0
3 -4.8500  2.4275  25.4952  10.4920  ...  23.0924  8.0280  3.2925  0
4 -3.8570  3.6650  25.1544  10.9456  ...  23.1008  8.0749  3.2350  0

[5 rows x 61 columns]

## Question 4
Why do we need normalization/standardization in this case?

In [47]:
# Print the maximum and the minimum of each each feature

### BEGIN SOLUTION
print("Feature wise maximum")
print(df.max())
print("Feature wise minimum")
print(df.min())
### END SOLUTION

Feature wise maximum
x1     -2.9435
x2      7.8475
x3     31.3416
x4     20.2336
x5      2.6190
        ...   
x57    16.2130
x58    23.9240
x59     8.2548
x60     4.0975
y       1.0000
Length: 61, dtype: float64
Feature wise minimum
x1     -4.9775
x2      2.0150
x3     24.0360
x4     10.1392
x5     -4.8727
        ...   
x57    16.0018
x58    23.0063
x59     8.0007
x60     3.0150
y       0.0000
Length: 61, dtype: float64


As we can see, the min and max of this dataset are quite close, but not as close as our previous dataset. So, We will normalize the dataset using preprocessing.minmax_scale() function.

In [48]:
# Use min-max scaling to normalize the features of X. Save the normalized array as X_norm. 
X_norm = preprocessing.minmax_scale(X) # REPLACE WITH APPROPRIATE CODE

# To confirm that it is now normalized, use the follwing code
df_norm = pd.DataFrame(X_norm, columns=df.columns[:-1])
df_norm.head()

### BEGIN SOLUTION

### END SOLUTION

x1        x2        x3  ...       x58       x59       x60
0  0.136431  0.156451  0.135677  ...  0.185355  0.245179  0.060046
1  0.323009  0.221603  0.272011  ...  0.105263  0.140496  0.087760
2  0.182153  0.246892  0.356110  ...  0.368421  0.258953  0.166282
3  0.062684  0.070724  0.199737  ...  0.093822  0.107438  0.256351
4  0.550885  0.282898  0.153088  ...  0.102975  0.292011  0.203233

[5 rows x 60 columns]

In [49]:
# Split the dataset (X_norm, y) into 85% training - 15% testing
# Then split the (X_train, y_train) again into 85% training and 15% validation.
# In both cases, use random_state=100
# Finally, print the shape of the X_train, X_valid, and X_test

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.15, random_state= 100) # replace with proper code
X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size= 0.15, random_state= 100) # replace with proper code

### BEGIN SOLUTION
print("Shape of X_train ={}".format(X_train.shape))
print("Shape of X_valid ={}".format(X_valid.shape))
print("Shape of X_test ={}".format(X_test.shape))
### END SOLUTION

Shape of X_train =(149, 60)
Shape of X_valid =(27, 60)
Shape of X_test =(32, 60)


## Quesion 5
We want to fit a Logistic Regression model. However, instead of manually setting the value of C (the inverse of the regularization strength), we want find the optimum value.
- On which data (train/test/valid) should we calculate the performance metric to find the optimum value of C?

To find the optimum value of C, we should calculate the performance state on validation data.

In [50]:
# Complete the follwing code to find the optimum value of C
C_candidate = [0.01, 0.1, 1, 10]
performance_metric_all = {}

for C in C_candidate:
    # define the Logistic Regression model with C = C and random_state=100
    # Fit the model using X_traina and y_train
    # predict the model's output for the ____ dataset (which one should it be?)
    # calculate the performance metric for the ____ dataset
    # save it in the array performance_metric_all by using performance_metric_all[C] = per...
    logistic_regression = linear_model.LogisticRegression(C = C, random_state= 100)
    logistic_regression.fit( X_train, y_train)
    y_pred_valid = logistic_regression.predict(X_valid)
    performance_metric_all[C] = accuracy_score(y_valid,y_pred_valid)
    pass

print(performance_metric_all)

### BEGIN SOLUTION

### END SOLUTION

{0.01: 0.6296296296296297, 0.1: 0.7037037037037037, 1: 0.6666666666666666, 10: 0.6666666666666666}


## Quesion 6
Based on the output of the previous block, which value of C should we choose?

We should choose C = 0.1 because it has the highest accuracy.


In [51]:
# Fit a logistic regression model on X_train, y_train using the C value found above. Use random_state=100
# predict the model's output for X_train and save it to y_predict_train
# predict the model's output for X_test and save it to y_predict_test

### BEGIN SOLUTION
logistic_regression = linear_model.LogisticRegression(C = 0.1, random_state = 100)
logistic_regression.fit(X_train, y_train)
y_pred_train =logistic_regression.predict(X_train)
y_pred_test = logistic_regression.predict(X_test)
### END SOLUTION

In [52]:
# Calculate a suitable performance metric for the training set for this classification problem
# Calculate a suitable performance metric for the test set for this classification problem
# Print the values

performance_train = accuracy_score(y_train, y_pred_train)
performance_test = accuracy_score(y_test, y_pred_test)

### BEGIN SOLUTION
print("The accuracy of train set: {}".format(performance_train))
print("The accuracy of test set: {}".format(performance_test))
### END SOLUTION

The accuracy of train set: 0.7986577181208053
The accuracy of test set: 0.875


## Question 7
- Which performance metric did you use and why?
- Is this a good model for the dataset? Why or why not?
- Suggest some ways (at least one) that might improve the performance metric

I used Sklearns accuracy score to determine the performance of the model. I used it because, it was a classification model and I want to know how much accurate our classifications were in training and test part of our datasets.

This is a quite good model for the dataset. Because it has 87.5% accuracy. But it can be improved to a better model.

We can improve the performance matric in a lot of way. Some of them are given below,

1. We can increase the amount of features.
2. We can use different Machine Learning or Deep Learning algorithms.
3. We can tweak the hyper parameters to get a better hyper parameter.
